### PR TITLE
Merge some lints together

### LINKED ISSUE #1078: Merge some lints together
Clippy has a lot of lints now, but sometimes there is not much difference between two lints. I think we should merge some of them together.
# Advantages
- Slightly shorter [big-list-of-all-the-lints](https://github.com/Manishearth/rust-clippy#lints).
- Less documentation duplication (see [`new_without_default`](https://github.com/Manishearth/rust-clippy/wiki#new_without_default) and [`new_without_default_derive`](https://github.com/Manishearth/rust-clippy/wiki#new_without_default_derive)).
- Two lints should be different only when one would want to set different allow/warn/deny level for them. Why would anyone ever want to forbid calls to `Option::unwrap` but not `Result::unwrap`?
- Less confusing as to why there are two different lints for that.
# Disadvantages
- Might be a breaking change, but IIRC `rustc` has a way to indicate a lint was renamed (and hopefully handle properly `#[allow]`s etc. on it).
- Need to find new names.
# Candidates

Here are some candidates (there might be other):
- [`block_in_if_condition_expr`](https://github.com/Manishearth/rust-clippy/wiki#block_in_if_condition_expr) and [`block_in_if_condition_stmt`](https://github.com/Manishearth/rust-clippy/wiki#block_in_if_condition_stmt) → rename them both to `block_in_if_condition`;
- [`option_map_unwrap_or`](https://github.com/Manishearth/rust-clippy/wiki#option_map_unwrap_or) and [`option_map_unwrap_or_else`](https://github.com/Manishearth/rust-clippy/wiki#option_map_unwrap_or_else) → `option_map_unwrap`;
- [`result_unwrap_used`](https://github.com/Manishearth/rust-clippy/wiki#result_unwrap_used) and [`option_unwrap_used`](https://github.com/Manishearth/rust-clippy/wiki#option_unwrap_used) → `unwrap_used`;
- [`for_loop_over_option`](https://github.com/Manishearth/rust-clippy/wiki#for_loop_over_option) and [`for_loop_over_result`](https://github.com/Manishearth/rust-clippy/wiki#for_loop_over_result) → ?
- [`new_without_default`](https://github.com/Manishearth/rust-clippy/wiki#new_without_default) and [`new_without_default_derive`](https://github.com/Manishearth/rust-clippy/wiki#new_without_default_derive) → `new_without_default` (the only difference is how the suggestion is built);
- [`string_add`](https://github.com/Manishearth/rust-clippy/wiki#string_add) and [`string_add_assign`](https://github.com/Manishearth/rust-clippy/wiki#string_add_assign) → `string_add`;
- [`identity_op`](https://github.com/Manishearth/rust-clippy/wiki#identity_op) and [`modulo_one`](https://github.com/Manishearth/rust-clippy/wiki#modulo_one) → `useless_arithmetic`

### PR DESCRIPTION
**Pull request by ThibsG**
This PR merges following lints:

- `block_in_if_condition_expr` and `block_in_if_condition_stmt` → `blocks_in_if_conditions`
- `option_map_unwrap_or`, `option_map_unwrap_or_else` and `result_map_unwrap_or_else` → `map_unwrap`
- `option_unwrap_used` and `result_unwrap_used` → `unwrap_used`
- `option_expect_used` and `result_expect_used` → `expect_used`
- `wrong_pub_self_convention` into `wrong_self_convention`
- `for_loop_over_option` and `for_loop_over_result` → `for_loops_over_fallibles`

Lints that have already been merged since the issue was created:
- [x] `new_without_default` and `new_without_default_derive` → `new_without_default`

Need more discussion:
- `string_add` and `string_add_assign`: do we agree to merge them or not? Is there something more to do? → **not merge finally**
- `identity_op` and `modulo_one` → `useless_arithmetic`: seems outdated, since `modulo_arithmetic` has been created.

fixes #1078

changelog: Merging some lints together: 
- `block_in_if_condition_expr` and `block_in_if_condition_stmt` → `blocks_in_if_conditions`
- `option_map_unwrap_or`, `option_map_unwrap_or_else` and `result_map_unwrap_or_else` → `map_unwrap_or`
- `option_unwrap_used` and `result_unwrap_used` → `unwrap_used`
- `option_expect_used` and `result_expect_used` → `expect_used`
- `for_loop_over_option` and `for_loop_over_result` → `for_loops_over_fallibles`

### CODE REVIEW

**Comment by ThibsG** [comment]
Btw, i'm not sure about `changelog` message, feel free to modify it.

**Comment by flip1995** [comment]
> * `block_in_if_condition_expr` and `block_in_if_condition_stmt` → `block_in_if_condition`

✔️ 

> * `option_map_unwrap_or`, `option_map_unwrap_or_else` and `result_map_unwrap_or_else` → `map_unwrap`

❌ →  `map_unwrap_or`

> * `option_unwrap_used` and `result_unwrap_used` → `unwrap_used`

✔️ 

> * `option_expect_used` and `result_expect_used` → `expect_used`

✔️ 

> * `wrong_pub_self_convention` into `wrong_self_convention`

❌  This is intentional, since the first one is a sylte lint and the second one is a restriction lint 

> * `for_loop_over_option` and `for_loop_over_result` → `for_loop_over_fallible`

✔️ 
 
> Need more discussion:
> 
> * `string_add` and `string_add_assign`: do we agree to merge them or not? Is there something more to do?

✔️ And move it to restriction

> * `identity_op` and `modulo_one` → `useless_arithmetic`: seems outdated, since `modulo_arithmetic` has been created.

❌ modulo_one is different from modulo_arithmetic (correctness vs restriction) also different intentions. I would also left `identity_op` separated, but would like some input on others for this one.

cc @rust-lang/clippy

**Comment by ThibsG** [comment]
> :x: → `map_unwrap_or`

Lint renamed.

> 
> > * `wrong_pub_self_convention` into `wrong_self_convention`
> 
> :x: This is intentional, since the first one is a sylte lint and the second one is a restriction lint

Commit deleted, lint is back.

> > * `string_add` and `string_add_assign`: do we agree to merge them or not? Is there something more to do?
> 
> :heavy_check_mark: And move it to restriction

Done

> 
> > * `identity_op` and `modulo_one` → `useless_arithmetic`: seems outdated, since `modulo_arithmetic` has been created.
> 
> :x: modulo_one is different from modulo_arithmetic (correctness vs restriction) also different intentions. I would also left `identity_op` separated, but would like some input on others for this one.
> 
> cc @rust-lang/clippy

No problem let's wait for feedback

**Comment by Manishearth** [comment]
@flip1995's suggestions seem good, though i'm a bit wary of the string_add one.

Also we should set these up as renamed lints so the old names still work and we have proper diagnostics.

**Comment by flip1995** [comment]
> though i'm a bit wary of the string_add one.

Oh I should have read the description of those lints more carefully. I thought one lints `x = x + s` and the other `x += s`. 

After reading it more carefully I would also keep them as separate lints. @ThibsG can you revert the commit for `string_add`? Sorry for the back and forth.

**Comment by ThibsG** [comment]
The `string_add` commit has been dropped and lints are now registered as renamed.

**Comment by ThibsG** [comment]
Do I need to change something in `CHANGELOG.md` ?

File: clippy_lints/src/block_in_if_condition.rs [hunk 1]
```diff
@@ -8,43 +8,40 @@ use rustc_middle::lint::in_external_macro;
...
-    /// if somefunc(|x| { x == 47 }) {}
+    /// // Bad
+    /// if { let x = somefunc(); x } { /* ... */ }
+    ///
+    /// // Good
+    /// let res = { let x = somefunc(); x };
+    /// if res { /* ... */ }
     /// ```
-    pub BLOCK_IN_IF_CONDITION_STMT,
+    pub BLOCK_IN_IF_CONDITION,
```
Comment (flip1995) [inline clippy_lints/src/block_in_if_condition.rs:39]: Let's rename them properly:

```suggestion
    pub BLOCKS_IN_IF_CONDITIONS,
```

File: clippy_lints/src/loops.rs [hunk 2]
```diff
@@ -168,55 +168,46 @@ declare_clippy_lint! {
...
-    /// ```ignore
-    /// if let Ok(x) = result {
-    ///     ..
+    /// // Good
+    /// if let Ok(x) = res {
+    ///     // ..
     /// }
     /// ```
-    pub FOR_LOOP_OVER_RESULT,
+    pub FOR_LOOP_OVER_FALLIBLE,
```
Comment (flip1995) [inline clippy_lints/src/loops.rs:208]:
```suggestion
    pub FOR_LOOPS_OVER_FALLIBLES,
```

**Comment by flip1995** [comment]
Thanks!

@bors r+

**Pull request merged**